fix: avoid conversion to bool in debug response

geeshta opened the pull request:
### Pull Request Checklist

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

- [x] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

### Description
Avoid implicit conversion to `bool` in `_debug_response.py` as this causes problems with some sqlalchemy classes, where instead of the original DB error being propagated, a `TypeError` is raised. See #2381 



### Close Issue(s)
Closes #2381

peterschutt commented on litestar/middleware/exceptions/_debug_response.py:
```diff
@@ -48,8 +48,13 @@ def get_symbol_name(frame: FrameInfo) -> str:
     # in instance and class methods
-    instance_or_cls = locals_dict.get("self") or locals_dict.get("cls")
-    classname = f"{get_name(instance_or_cls)}." if instance_or_cls else ""
+    instance_or_cls = None
+    if "self" in locals_dict:
+        instance_or_cls = locals_dict["self"]
+    elif "cls" in locals_dict:
+        instance_or_cls = locals_dict["cls"]
```
```suggestion
    instance_or_cls = inst if (inst := locals_dict.get("self")) is not None else locals_dict.get("cls")
```
Can still be done in compact form - just a nit-pick though, these are functionally the same.

peterschutt reviewed (changes requested):
I suggest adding a test - this is the type of thing that a tool, or an overzealous refactor might reverse at some point down the track without realizing why it is why it is.

This reproduces it:

```py
from inspect import getinnerframes

from litestar.middleware.exceptions._debug_response import get_symbol_name


def test_get_symbol_name_where_type_doesnt_support_bool() -> None:
    class Test:
        def __bool__(self) -> bool:
            raise TypeError("This type doesn't support bool")

        def method(self) -> None:
            raise RuntimeError("Oh no!")

    try:
        Test().method()
    except Exception as e:
        exc = e

    frame = getinnerframes(exc.__traceback__, 2)[-1]
    assert get_symbol_name(frame) == "Test.method"
```

geeshta commented:
> I suggest adding a test - this is the type of thing that a tool, or an overzealous refactor might reverse at some point down the track without realizing why it is why it is.
> 
> This reproduces it:
> 
> ```python
> from inspect import getinnerframes
> 
> from litestar.middleware.exceptions._debug_response import get_symbol_name
> 
> 
> def test_get_symbol_name_where_type_doesnt_support_bool() -> None:
>     class Test:
>         def __bool__(self) -> bool:
>             raise TypeError("This type doesn't support bool")
> 
>         def method(self) -> None:
>             raise RuntimeError("Oh no!")
> 
>     try:
>         Test().method()
>     except Exception as e:
>         exc = e
> 
>     frame = getinnerframes(exc.__traceback__, 2)[-1]
>     assert get_symbol_name(frame) == "Test.method"
> ```

I've added the test. I added some type guards because pyright complained about `exc` being potentially unbound and `__traceback__` being potentially None... other solution would by just `# type: ignore` them.

peterschutt reviewed (approved):
Thanks for this!

peterschutt commented:
@all-contributors add @geeshta for code bug
